merging gilt repo changes + my fixes and updates

patsissons opened the pull request:
Let me know if you have any issues with my suggested changes. I'm pretty open to discuss any of the decisions I have made.

webschik commented on package.json:
```diff
@@ -9,35 +9,47 @@
     "less"
   ],
-  "main": "dist/less-syntax.js",
-  "scripts": {
```
@patsissons, why not NPM scripts?

webschik commented on .jscsrc:
```diff
@@ -1,40 +0,0 @@
-{
```
@patsissons, we can't remove this file. Also as a validation process. I added this is no accident, JSCS helps to have one code style and write code that's based one 1 style guide. I think we can't remove that! Now I see different variables naming, maybe there are another problems with code style.

patsissons commented:
removed the `jscs` file because it was redundant (with `eslint`). See the `eslint` [README](https://github.com/eslint/eslint#how-does-eslint-performance-compare-to-jshint-and-jscs) for the reasoning.

The npm scripts are no longer necessary given the `gulp` build configuration. `gulp` essentially augments each of those npm scripts with proper task dependency structure.

webschik commented:
@patsissons , but ESLint doesn't cover all code style issues. Look, now we have 2 different naming convention in vars, in future we will have to review every MR by themselves to check a single code style - JSCS does the same things and makes our life easier :) 

I see Gulp, but we don't really need it and I try to not use excess tools. NPM script does the same work as Gulp now, but without overhead: we just need to call some tools (Mocha, JSCS, ESLint)

patsissons commented:
I will replace `eslint:recommended` with [strict](https://www.npmjs.com/package/eslint-config-strict) to bring more code format and style consistency. Otherwise I think `ESLint` should cover any style rule you are looking for (if you find that it isn't please let me know which rule you would like to see covered).

I choose to use `Gulp` mostly because of it's simplicity and power, but additionally because of the local continuous integration potential (i.e. `gulp.watch(...)`). This is something that npm scripts simply cannot reproduce without a proper build tool.

webschik commented:
@patsissons, I agree about `gulp.watch`, okay! Also I would ask you for what we need `es6-shim`?

About code style (JSCS rules):
- [validateIndentation](http://jscs.info/rule/validateIndentation) (I, and many devs, always use 4 spaces, why should I change it?)
- [disallowSpacesInsideBrackets](http://jscs.info/rule/disallowSpacesInsideBrackets)
- [disallowSpacesInsideArrayBrackets](http://jscs.info/rule/disallowSpacesInsideArrayBrackets)
- [requireSpacesInFunction](http://jscs.info/rule/requireSpacesInFunction)
- [requireCurlyBraces](http://jscs.info/rule/requireCurlyBraces)
- [requirePaddingNewLinesBeforeLineComments](http://jscs.info/rule/requirePaddingNewLinesBeforeLineComments)
- [disallowTrailingComma](http://jscs.info/rule/disallowTrailingComma)

Guys, I understand that we have your view on the code style, another habits, but I as a creator of this library also approved some code style and code validation, added the tools like ESLint & JSCS to help somebody with contributions to this project. Now you propose to break down all these rules and start from beginning. I had another thoughts about contribution...

shellscape commented:
imho this PR tries to do too much.

31 commits is way to many for a PR to this lib, too much to ask the owner to review in a timely manner. They should have been squashed, or incrementally PR'd.
The changes to code should have been one PR.
The changes to meta files should have been a separate PR.
The changes to tooling should have been a separate PR.

The .vscode directory shouldn't be in the PR.
The .editorconfig file's worth is a religious debate and I don't think it should be in the PR. 

@webschik you're well within your right to reject it because of the tooling changes. as the owner of the repo it's your choice to setup contribution guidelines, and I would recommend creating a CONTRIBUTING.md file so that other developers can reference them. 

That being said, I do agree with the use of Gulp (postcss-scss uses it well, and what's being proposed is just copy of that). The code changes are good, and could be cherry picked.

shellscape commented on LICENSE:
```diff
@@ -1,6 +1,8 @@
 
+Copyright (c) 2013 Andrey Sitnik <andrey@sitnik.ru>
 Copyright (c) 2016 Denys
+Copyright (c) 2016 Pat Sissons <patricksissons@gmail.com>
```
Copyright in MIT is basically pointless past calling out the initial developer. MIT is a free and clear license. Generally considered poor form to add to it, as you'll show in the contributors. It's up to the repo own wether or not to have a contributors list.

shellscape commented on README.md:
```diff
@@ -1,31 +1,70 @@
-# PostCSS LESS syntax
```
Good changes to the readme, but should have been a separate PR imho.

webschik commented:
@patsissons  It makes sense to separate this PR to small ones. 
Also I created a small [contribution guide](https://github.com/webschik/postcss-less/blob/master/CONTRIBUTING.md).

patsissons commented:
Hey guys,

thanks for the comments and feedback. I'll try and respond to each item with a defense or reasoning for the choice I made (in no particular order).

This PR is indeed massive, but expectedly so. I am merging 2 (technically 3, if you count my original work that was never published to github) independent efforts to accomplish the same goal. Some files lined up, some didn't. Some files were easy to merge, others not so much. 

All three of us had very different styles, so I thought the best approach would be to choose a common neutral `ESLint` shareable config ([strict](https://github.com/keithamus/eslint-config-strict)). I chose strict because it would be hard to disagree with its rule choices (since it's purpose for existing was a common strict standard). I then applied changes to the 3 code bases to adhere to this strict standard of style.

I would have liked to do incremental PR's but at the initial time of embarking on this _3 way merge_ effort I was unaware of how large and complex this would become. Because of that, the merges are now so intertwined that it is a little challenging to split up this PR. That is not to say I cannot squash and split some of the independent changes, but there is at the very least going to be one fairly large PR (large as in unified changes).

Since the tooling and meta file changes (adds and modifications) are a large part of the reasoning behind many other changes it doesn't make sense to submit a PR for (most of) them separately. Though I agree some of those files could probably reside in their own PR (I'll look into this further when I attempt to split this PR up).

The `.vscode` directory is equivalent to the other meta files that define the development environment. You could perhaps argue that the `launch.json` and `tasks.json` don't need to belong, but you could just as easily argue that they do. The `settings.json` file is an absolute must as it defines workspace environment settings. Similarly, I included a `.editorconfig` which is an attempt to keep the workspace environment identical regardless of choice of editing tools (though I don't actually use any tools that support a `.editorconfig` properly at the moment). Essentially I included this file for completeness and consistency.

As for the license file, as I mentioned earlier I was merging 3 independent efforts each with their own (MIT) license file. During merging I simply merged the three together. I'm not sure what is appropriate in this situation and I'm perfectly content removing my attribution on the license as my original contributions to this code base are now _almost_ exclusively *augmentation* and not creative design. What do you propose for the attribution from the `gilt` repo? I recall reading a stackoverflow thread about merging the licenses of two open source projects and following the best answer (which was to just merge attribution). Can't seem to find the link anymore, so you'll have to trust me that I based my choice off that answer. I don't really care if my name is on it or not, so removing it is perfectly acceptable to me.

Finally, as for linting/style checking. I removed JSCS because ESLint (afaik) can capture both linting and style enforcing (which is why `eslint-config-strict` impacted so many style changes). My reasoning for all the changes was simply two adhere to *SOME* standard style between all efforts. I think I accomplished this by only overriding a single rule from the `strict` ruleset and a handful of `next-line` overrides in the source. Every override applied is justified in the comments. This approach, I believe, is fair and consistent across multiple efforts (for context it was not aligned with any of the 3 efforts being merged, even my own). That said, for reasons of repo ownership you could easily argue that it is legitimate to override a rule like `indent` or `lines-around-comment` which is purely a project style decision. However, some of the other rules might be worth a discussion. I would propose discussing the merits of the following:
* `disallowSpacesInsideBrackets` (`object-curly-spacing` in `eslint`, enforced in `strict`)
* `disallowSpacesInsideArrayBrackets` (`array-bracket-spacing` in `eslint`, enforced in `strict`)
* `requireSpacesInFunction` (`no-extra-parens` in `eslint`, enforced for functions in `strict`)
* `disallowTrailingComma` (`comma-dangle` in `eslint`, enforced for multiline items in `strict`)
I believe `requireCurlyBraces` (`curly` in `eslint`, enforced for all control statements) is already aligned. Ultimately, it is your repo and your decision, I just wanted to justify that I was not trying to usurp your style but rather conform to an existing strict style. To break away from any rule in the strict style there should be some justification (in the form of discussion). I believe it is clear however, that `jscs` is redundant, as `eslint` captures all of the rules you have mentioned, please correct me if I'm wrong.
 
Sorry for the wall of text. I'm definitely not trying to fight for my changes just provide visibility on the reasoning for my choices. I'm still completely open to any changes as ultimately my goal is to just bring `postcss-less` support to `stylelint`. I'm not sure when I'll get a chance to look at cutting down this PR into multiple smaller ones (one would hope for some time this week, but I have another project that needs work this week as well). I don't think there is any immediate need to get on this but if anyone is truly waiting for this merge I can perhaps prioritize my evening efforts.

webschik commented:
@patsissons , thanks you a lot! I don't want to impose my view too and would like to make our interaction easier. What I propose:
- I will get this PR, fix some issues according to JSCS config
- .vscode I'm going to add in .gitignore, like .idea file, because not all use VSCode (for example I'm using WebStorm)
- I'm going to accept this PR, because all tests are passed
- We will do next PRs smaller and will separated new features by each PR.
- For next contributors there is a Contribution guide with description about our rules and tools.

patsissons commented:
If you want any help configuring and reversing any of the style rules let me know. Or just any other help in general cleaning up this PR.

shellscape commented:
cheers for open source.

I think I'll continue on with the gilt effort separately. We were considering abandoning it and moving forward with this effort, but this PR adds far too much cruft for my tastes. Instead, we'll cherry pick the additions and add credit where it is due. Please feel free to do the same if you notice improvements not already contained here.

webschik commented:
@shellscape, I will check and make some fixes in this PR for code style and excess files/folders ASAP, so stay in touch.

webschik commented:
@patsissons, I've pull this PR and refactored the code, fixed some configuration in build files. Should I create a new PR for it? Because I don't have an access to push to your branch `patsissons-gilt-merge` .

patsissons commented:
I think it probably makes most sense to create a new PR. But up to you, I can grant push access to the repo in a sec.

webschik commented:
@patsissons, thank you! I'm going to merge PR today.

patsissons commented:
great news! we should have full support in postcss very soon. I may add some additional tests later if postcss tests are still failing after you publish to npm.

The current failing test is `h1 { max-height: #{($line-height) * ($lines-to-show)}em; }` (`CssSyntaxError: <css input>:1:35: Unknown word`).

webschik commented:
@patsissons, but it's not a valid LESS code, please, [check this](http://lesscss.org/less-preview/#%7B%22less%22%3A%22h1%20%7B%20%5Cn%20%20max-height%3A%20%23%7B(%24line-height)%20*%20(%24lines-to-show)%7Dem%3B%20%5Cn%7D%22%7D) and [this](http://lesscss.org/less-preview/#%7B%22less%22%3A%22h1%20%7B%20%5Cn%20%20%40line-height%3A%201em%3B%5Cn%20%20%40lines-to-show%3A%202%3B%5Cn%20%20%5Cn%20%20max-height%3A%20%40line-height%20*%20%40lines-to-show%3B%20%5Cn%7D%22%7D)

patsissons commented:
yea good point i need to go through my PR there and ensure the new less tests are valid (since i just copied them from the scss tests).

patsissons commented:
I updated the stylelint tests so that all tested snippets were the equivalent valid less (where applicable, the `map-get` one has no valid less translation). Only one test is not currently passing and it is actually on the same test I posted above. It may be a legit lexing/parsing issue.

```
css: "h1 { max-height: ((@line-height) * (@lines-to-show))em; }"
rule: function-whitespace-after
options: "always"
# warning: Expected whitespace after ")" (function-whitespace-after)
not ok 6020 LESS-style interpolation should pass
  ---
    operator: equal
    expected: 0
    actual:   1
  ...
```

thoughts?

webschik commented:
@patsissons, I will check this bug with PR. If it doesn't fix the bug, I will try to resolve this problem.

webschik commented:
@patsissons, I will check it in PR and try to resolve this bug.

webschik commented:
@patsissons, the problem is with inner parenthesis. I added this test case and updated the PR. Give me some time to fix that and we will be able to merge this PR. I hope :)

webschik commented:
@patsissons, I've fixed that and merged PR. All tests are passed.
